compiler/rustc_codegen_llvm/src/back/lto.rs

```diff
@@ -664,6 +665,30 @@ pub(crate) fn run_pass_manager(
             unsafe { llvm::LLVMDumpModule(module.module_llvm.llmod()) };
         }
 
+        let cx =
+            SimpleCx::new(module.module_llvm.llmod(), &module.module_llvm.llcx, cgcx.pointer_size);
+
+        for function in cx.get_functions() {
+            let name = get_value_name(function);
+            let name = std::str::from_utf8(name).unwrap();
+
+            if name.starts_with("__enzyme") {
+                // Ensure `noinline` is present before replacing it.
+                // This is not strictly necessary for correctness, but serves as a sanity check
+                // in case the autodiff pass stops injecting `noinline` in the future.
+                assert!(
+                    !attributes::has_attr(function, 0, llvm::AttributeKind::NoInline),
```

Comment:
If you look at the AttributePlace definition, you'll see that 0 means return attribute, if you want a function it has to be !0. I'd recommend to just use the AttributePlace, like in your other locations. Still investigating why the attribute isn't removed.